source/common/router/config_impl.cc

```diff
@@ -71,20 +71,27 @@ ShadowPolicyImpl::ShadowPolicyImpl(const envoy::api::v2::RouteAction& config) {
 
 HashPolicyImpl::HashPolicyImpl(
     const Protobuf::RepeatedPtrField<envoy::api::v2::RouteAction::HashPolicy>& hash_policy)
-    : header_name_(hash_policy[0].header().header_name()) {
+    : header_name_(hash_policy[0].header().header_name()),
+      hash_ip_(hash_policy[0].connection_properties().source_ip()) {
   // TODO(htuch): Add support for multiple hash policies, #1422.
   ASSERT(hash_policy.size() == 1);
-  // TODO(htuch): Add support for cookie and source IP hash policies, #1295 and
-  // #1296.
+  // TODO(htuch): Add support for cookie hash policies, #1295
   ASSERT(hash_policy[0].policy_specifier_case() ==
-         envoy::api::v2::RouteAction::HashPolicy::kHeader);
+             envoy::api::v2::RouteAction::HashPolicy::kHeader ||
+         hash_policy[0].policy_specifier_case() ==
+             envoy::api::v2::RouteAction::HashPolicy::kConnectionProperties);
 }
 
-Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers) const {
+Optional<uint64_t> HashPolicyImpl::generateHash(const std::string& downstream_addr,
```

Comment:
According to @mattklein123:

```IIRC AsyncClient does not have a downstream connection. This actually confused me also and I had to think about it so I think worth a comment.```

Should we at least add a check that a downstream IP based route config is not used alongside this?

Reply:
I don't understand your question, but a comment was added to the docs for generateHash that addresses @mattklein123's comment. `downstream_addr` is an empty string for requests originated by AsyncClient.

Reply:
OK, as long as it's empty it's fine.

Reply:
OK cool. I wasn't sure. If it's empty string that's fine.